Replace Eclipse references with VSCode [PAB-4814] (#3357)
SoftwareAG/c8y-docs: content/streaming-analytics/epl-apps-bundle/basic-functionality.md

### Developing apps {#developing-apps}

> An EPL app is a Cumulocity application written in a single EPL (\*.mon) file. You can develop EPL apps in two different ways:

Reviewer: ```suggestion
An EPL app is a {{< product-c8y-iot >}} application written in a single EPL (\*.mon) file. You can develop EPL apps in two different ways:
```

* You can use the [Streaming Analytics application](#dev-apps-with-sa) which is available from {{< product-c8y-iot >}}'s application switcher and develop your EPL apps within {{< product-c8y-iot >}}.

> * Or you can create a Git repository for your EPL apps and develop it on your local machine with the [Apama Extension](#apama-local-dev) for Microsoft Visual Studio Code. This approach also works for larger EPL applications deployed to a custom microservice. 

Reviewer: ```suggestion
* Or you can create a Git repository for your EPL apps and develop it on your local machine with the [Apama extension](#apama-local-dev) for Microsoft Visual Studio Code. This approach also works for larger EPL apps deployed to a custom microservice. 
```

Author: @BeateRixen I'd prefer not to make this particular change:
a) We always refer to it as the "Apama Extension for Microsoft Visual Studio Code", its a "proper name"... not just any old "extension" but The Apama Extension,
b) I explicitly want to make clear this is NOT just for EPL apps (= small, single-file EPL applications) but also for larger EPL applications (i.e. _not_ EPL apps).

Reviewer: @ben-c8y 
a) Below you use "Apama extension" which is why I changed it here to be consistent. Please check.
b) Got it, thanks for the explanation.

Author: Thanks @BeateRixen OK I fixed the other references to say "Apama Extension"

See also [Deploying apps](/streaming-analytics/epl-apps/#deploying-apps).

For information about a more systematic and automated approach to testing EPL apps, see [Testing apps](/streaming-analytics/epl-apps/#testing-apps).

#### Developing apps with the {{< apama-vscode >}} {#apama-local-dev}

The {{< apama-vscode >}} provides a full development environment and is the tool of choice when you have a complex EPL application. When your EPL app (that is, the \*.mon file) is ready, you must import (deploy) it into {{< product-c8y-iot >}}.

##### Step 1 - Install the {{< apama-vscode >}} {#step-1---install-apama-vscode}

1. **Install Visual Studio Code**: Download and install [Microsoft Visual Studio Code](https://code.visualstudio.com/>).

> 2. **Install Apama Extension**: Install the [Apama extension](https://marketplace.visualstudio.com/items?itemName=ApamaCommunity.apama-extensions>) from the Visual Studio Code Marketplace.

Reviewer: See here.

Author: Fixed

3. Following the steps listed on the extension page to setup WSL (if using Windows) and to install a container engine for running Apama inside a container, or else install Apama locally.

##### Step 2 - Create a project and repository {#step-2---create-a-project}

> The best way to start EPL Apps development is to create a new Git repository based on the 

Reviewer: ```suggestion
The best way to start EPL apps development is to create a new Git repository based on the 
```

[Streaming Analytics Sample Repository Template](https://github.com/Cumulocity-IoT/streaming-analytics-sample-repo-template). 

> Go to that page in GitHub, and click the button to "Use this template" and then "Create a new repository". 

Reviewer: ```suggestion
Go to that page in GitHub, and click **Use this template** and then **Create a new repository**. 
```

Projects created from the template already have the main bundles you will need for EPL apps. 

If you already have a Git repository and Apama project for your application, just copy across the `.devcontainer` directory from the template repository 
so that it can be opened in a VS Code Dev Container. 

##### Step 3 - Open the repository in VS Code {#step-3---open-project}

To open your Git repository in VS Code, open the VS Code command palette (`F1`), run `Dev Containers: Clone Repository in Container Volume` 
and then enter the `https://` link to your GitHub repository. This assumes you have a Docker-compliant container engine installed; if not see the 
[Apama extension](https://marketplace.visualstudio.com/items?itemName=ApamaCommunity.apama-extensions>) documentation for 
information about how to work with project folders using a local installation of Apama. 

Instead of using VS Code you can use a web browser to open the repository in [GitHub Codespaces](https://github.com/features/codespaces) without installing anything at all. 

##### Step 4 - Create an EPL file for your application {#step-4---create-a-monitor-file}

> If using the template repository, you will find a sample EPL app in the `src/` directory of your repository. This is a good starting point for your own EPL apps. Feel free to customize that file, or replace it with your own application, perhaps based on the other samples in the [EPL Apps Tools](https://github.com/Cumulocity-IoT/apama-eplapps-tools) repository. You can also create additional `.mon` files, one for each EPL app. 

Reviewer: ```suggestion
If using the template repository, you will find a sample EPL app in the *src/* directory of your repository. This is a good starting point for your own EPL apps. Feel free to customize that file, or replace it with your own application, perhaps based on the other samples in the [EPL Apps Tools](https://github.com/Cumulocity-IoT/apama-eplapps-tools) repository. You can also create additional *.mon* files, one for each EPL app. 
```

> For more information about interacting with Cumulocity from your application, see [The Cumulocity Transport Connectivity Plug-in]({{< link-apama-webhelp >}}/standard-connectivity-plugins/the-cumulocity-iot-transport-connectivity-plug-in/) in the Apama documentation.

Reviewer: ```suggestion
For more information about interacting with {{< product-c8y-iot >}} from your application, see [The Cumulocity Transport Connectivity Plug-in]({{< link-apama-webhelp >}}/standard-connectivity-plugins/the-cumulocity-iot-transport-connectivity-plug-in/) in the Apama documentation.
```

##### Step 5 - Test the EPL {#step-5---run-and-test-the-monitor-file}

Before you import the newly created file as an EPL app into {{< product-c8y-iot >}} and activate it there, you might want to test if the file works as expected from within the {{< apama-vscode >}}. 

> First, check the "Problems" view for any basic errors such as incorrect syntax. Then, create a [test](/streaming-analytics/epl-apps/#testing-apps) that runs your application in a local correlator. If you used the template to create your repository you will find a sample test to get started under the `tests/` directory. 

Reviewer: ```suggestion
First, check the "Problems" view for any basic errors such as incorrect syntax. Then, create a [test](/streaming-analytics/epl-apps/#testing-apps) that runs your application in a local correlator. If you used the template to create your repository you will find a sample test to get started under the *tests/* directory. 
```

##### To deploy an Apama application as a microservice {#to-deploy-an-apama-application-as-a-microservice}

1. Develop your application in the {{< apama-vscode >}}.

> 2. Add a `Dockerfile` as described by [Deploying Apama Applications with Docker]({{< link-apama-webhelp >}}/deploying-and-managing-apama-applications/overview-of-deploying-apama-applications/). The sample file in the `Apama/etc/Dockerfile.project` directory of the Apama installation is a good starting point. 

Reviewer: ```suggestion
2. Add a Dockerfile as described by [Deploying Apama Applications with Docker]({{< link-apama-webhelp >}}/deploying-and-managing-apama-applications/overview-of-deploying-apama-applications/). The sample file in the *Apama/etc/Dockerfile.project* directory of the Apama installation is a good starting point. 
```

If you still use the images from the previous location, you must migrate them.
{{< /c8y-admon-info >}}

### Testing apps {#testing-apps}

The best way to check your EPL applications behave correctly is to create automated tests for them. 

> The [Apama EPL Apps Tools](https://github.com/Cumulocity-IoT/apama-eplapps-tools) GitHub repository includes extensions for the PySys test framework that allow you to simply write tests for your EPL apps and to run them automatically, either by uploading the EPL to Cumulocity, or by running inside a local correlator communicating with Cumulocity. See the [EPL Apps Tools documentation](https://cumulocity-iot.github.io/apama-eplapps-tools/) for detailed information. For more information on PySys, see the [API Reference for Python]({{< link-apama-webhelp >}}/related/pydoc/index.html) in the main Apama documentation.

Reviewer: ```suggestion
The [Apama EPL Apps Tools](https://github.com/Cumulocity-IoT/apama-eplapps-tools) GitHub repository includes extensions for the PySys test framework that allow you to simply write tests for your EPL apps and to run them automatically, either by uploading the EPL to {{< product-c8y-iot >}}, or by running inside a local correlator communicating with {{< product-c8y-iot >}}. See the [EPL Apps Tools documentation](https://cumulocity-iot.github.io/apama-eplapps-tools/) for detailed information. For more information on PySys, see the [API Reference for Python]({{< link-apama-webhelp >}}/related/pydoc/index.html) in the main Apama documentation.
```

> To provide the credentials and server URL for your Cumulocity tenant, you need to set some environment variables when running the tests:

Reviewer: ```suggestion
To provide the credentials and server URL for your {{< product-c8y-iot >}} tenant, you need to set some environment variables when running the tests:
```

```
CUMULOCITY_SERVER_URL=<URL>
CUMULOCITY_USERNAME=<USERNAME>
CUMULOCITY_PASSWORD=<PASSWORD>
```

As well as running tests locally during development, you can script deployment and testing of your EPL apps for CI/CD (continuous integration and continuous delivery) purposes.

> ### Connectivity bundles for communicating with Cumulocity

Reviewer: ```suggestion
### Connectivity bundles for communicating with Cumulocity {#connectivity-bundles}
```

> The **Cumulocity IoT > Cumulocity Notifications 2.0** bundle exposes the {{< product-c8y-iot >}} client to EPL apps using the Notifications 2.0 mechanism.

Reviewer: Should be "Cumulocity" instead of Cumulocity IoT; not sure what is the actual name though

Author: Yes ideally we'd change all IoT references in the product but that's a big job as there are LOADS of IoT references, and a lot of testcases that would break if we changed things, and doc to update in a large set of repos... and anyway that's out of scope of this PR
